src/bandersnatch_storage_plugins/s3.py

```diff
@@ -0,0 +1,368 @@
+        if path is None:
+            path = str(self.mirror_base_path / ".lock")
+        return S3FileLock(path, backend=self)
+
+    def walk(self, root: PATH_TYPES, dirs: bool = True) -> List[S3Path]:
+        if not isinstance(root, self.PATH_BACKEND):
+            root = self.PATH_BACKEND(root)
+
+        results: List[S3Path] = []
+        for pth in root.iterdir():
+            if pth.is_dir():
+                if dirs:
+                    results.append(pth)
+                for subpath in self.walk(pth, dirs=dirs):
+                    results.append(pth / subpath)
+            else:
+                results.append(pth)
+        return results
+
+    def find(self, root: PATH_TYPES, dirs: bool = True) -> str:
+        """Really discourage using this method, this will take a lot of time
+        S3Path.glob or S3Path.rglob will be better
+        """
+        results = self.walk(root, dirs=dirs)
+        results.sort()
+        return "\n".join(str(result.relative_to(root)) for result in results)
+
+    def rewrite(
+            self, filepath: PATH_TYPES, mode: str = "w", **kw: Any
+    ) -> IO:
+        """Rewrite an existing file atomically to avoid programs running in
+        parallel to have race conditions while reading."""
+        if not isinstance(filepath, self.PATH_BACKEND):
+            filepath = self.PATH_BACKEND(filepath)
+        return filepath.open(mode=mode, **kw)
+
+    @contextlib.contextmanager
+    def update_safe(self, filename: PATH_TYPES, **kw: Any) -> IO:
+        """Rewrite a file atomically.
+
+        Clients are allowed to delete the tmpfile to signal that they don't
+        want to have it updated.
+        """
+        with tempfile.NamedTemporaryFile(
+                delete=False,
+                prefix=f"{os.path.basename(filename)}.",
+                **kw,
+        ) as tf:
+            tf.has_changed = False  # type: ignore
+            yield tf
+            if not os.path.exists(tf.name):
+                return
+            local_filename_tmp = pathlib.Path(tf.name)
+        self.copy_local_file(str(local_filename_tmp), str(filename))
+        local_filename_tmp.unlink()
+
+    def copy_local_file(self, source: PATH_TYPES, dest: PATH_TYPES) -> None:
+        """Copy the contents of a local file to a destination in swift"""
+        with open(source, "rb") as fh:
+            self.write_file(str(dest), fh.read())
+        return
+
+    def compare_files(self, file1: PATH_TYPES, file2: PATH_TYPES) -> bool:
+        """Compare two files, returning true if they are the same and False if not."""
+        file1_contents = self.read_file(file1, text=False)
+        file2_contents = self.read_file(file2, text=False)
+        assert isinstance(file1_contents, bytes)
+        assert isinstance(file2_contents, bytes)
+        file1_hash = hashlib.sha256(file1_contents).hexdigest()
+        file2_hash = hashlib.sha256(file2_contents).hexdigest()
+        return file1_hash == file2_hash
```

Comment:
Should we read cunt by chunk to cater for big files? If filesystem non swift do it, this is fine, but we should open an issue. there are some big files in PyPI these days ...

Reply:
I found that he s3 provide a way to [get file MD5 without downloading it](https://stackoverflow.com/questions/26415923/boto-get-md5-s3-file). maybe we can use this if two files are both on s3. 

Should I consider comparing two files in two kinds of filesystem?